src/Symfony/Component/Validator/Constraints/BackedEnumValueValidator.php

```diff
@@ -0,0 +1,79 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+/**
+ * BackedEnumValueValidator validates that the value is one of the expected values.
+ *
+ * @author Aurélien Pillevesse <aurelienpillevesse@hotmail.fr>
+ */
+class BackedEnumValueValidator extends ConstraintValidator
+{
+    public function validate(mixed $value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof BackedEnumValue) {
+            throw new UnexpectedTypeException($constraint, BackedEnumValue::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        try {
+            $enumTypeValue = $constraint->type::tryFrom($value);
+        } catch (\TypeError) {
+            $this->context->buildViolation($constraint->typeMessage)
+                ->setParameter('{{ type }}', $this->formatValue((string) (new \ReflectionEnum($constraint->type))->getBackingType()))
+                ->setCode(BackedEnumValue::INVALID_TYPE_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (null === $enumTypeValue) {
+            $values = $constraint->type::cases();
+
+            if (\count($constraint->except) > 0) {
+                $values = $this->casesFromBackedEnumWithoutExceptOnes($values, $constraint->except);
+            }
+
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setParameter('{{ choices }}', $this->formatValues(array_map(static fn (\BackedEnum $case) => $case->value, $values)))
+                ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (\count($constraint->except) > 0 && \in_array($enumTypeValue, $constraint->except, true)) {
+            $values = $this->casesFromBackedEnumWithoutExceptOnes($constraint->type::cases(), $constraint->except);
+
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($enumTypeValue->value))
+                ->setParameter('{{ choices }}', $this->formatValues(array_map(static fn (\BackedEnum $case) => $case->value, $values)))
+                ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function casesFromBackedEnumWithoutExceptOnes(array $cases, array $except): array
+    {
+        return array_filter($cases, function ($currentValue) use ($except) {
+            return !\in_array($currentValue, $except, true);
+        });
+    }
```

Comment:
There's more duplicated logic that can be extracted into this method.

```suggestion
    private function formatValidCases(BackedEnumValue $constraint): string
    {
        return $this->formatValues(array_map(
            static fn (\BackedEnum $case) => $case->value,
            array_filter(
                $constraint->type::cases(),
                static fn ($currentValue) => !\in_array($currentValue, $constraint->except, true),
            ),
        ));
    }
```

Reply:
Fixed, sorry for all the feedbacks :sweat_smile: